zcash_primitives/src/transaction/components/amount.rs

```diff
@@ -0,0 +1,204 @@
+use std::iter::Sum;
+use std::ops::{Add, AddAssign, Sub, SubAssign};
+
+const COIN: i64 = 1_0000_0000;
+const MAX_MONEY: i64 = 21_000_000 * COIN;
+
+pub const DEFAULT_FEE: Amount = Amount(10000);
+
+/// A type-safe representation of some quantity of Zcash.
+///
+/// An Amount can only be constructed from an integer that is within the valid monetary
+/// range of `{-MAX_MONEY..MAX_MONEY}` (where `MAX_MONEY` = 21,000,000 × 10⁸ zatoshis).
+/// However, this range is not preserved as an invariant internally; it is possible to
+/// add two valid Amounts together to obtain an invalid Amount. It is the user's
+/// responsibility to handle the result of serializing potentially-invalid Amounts. In
+/// particular, a [`Transaction`] containing serialized invalid Amounts will be rejected
+/// by the network consensus rules.
+///
+/// [`Transaction`]: crate::transaction::Transaction
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct Amount(i64);
+
+impl Amount {
+    /// Returns a zero-valued Amount.
+    pub const fn zero() -> Self {
+        Amount(0)
+    }
+
+    /// Creates an Amount from an i64.
+    ///
+    /// Returns an error if the amount is outside the range `{-MAX_MONEY..MAX_MONEY}`.
+    pub fn from_i64(amount: i64) -> Result<Self, ()> {
```

Comment:
Would it make sense to change the return type here and for similar functions from `Result<Self, ()>` to `Option<Self>`?

Reply:
I think I prefer this being an explicit (if unit) error. I know we are inconsistent about this in our API; we should open an issue for going through and refactoring everything to be internally consistent in how we decided to return errors vs `None`.

Reply:
It would be great to go through and make these consistent, but I am happy leaving this for now.